### PR TITLE
Feature/transform

### PR DESCRIPTION
**Pull request by dzhwinter**
add Datatype, Layout transform.

### CODE REVIEW

File: paddle/framework/data_transform.h [hunk 1]
```diff
@@ -45,6 +48,42 @@ struct KernelTypePairHash {
   }
 };
 
+template <typename InType, typename OutType>
+struct CastDataTypeFunctor {
+  HOSTDEVICE OutType operator()(InType in) const {
```
Comment (chengduoZH) [inline paddle/framework/data_transform.h]: Add `inline`

File: paddle/framework/data_transform.h [hunk 2]
```diff
@@ -45,6 +48,42 @@ struct KernelTypePairHash {
...
+    auto* in_end = in_begin + numel;
+    auto* out_begin = out_->mutable_data<OutType>(place);
+    if (platform::is_cpu_place(place)) {
+      platform::Transform<platform::CPUDeviceContext> trans;
+      auto* context = static_cast<const platform::CPUDeviceContext*>(ctx_);
+      trans(*context, in_begin, in_end, out_begin,
+            CastDataTypeFunctor<InType, OutType>());
+    } else {
+      // TODO(dzhwinter): enhance CopyFrom CPU<->GPU with different data type?
+      PADDLE_THROW("Unsupport CPU <-> GPU!");
```
Comment (chengduoZH) [inline paddle/framework/data_transform.h]: `OpKernelType` is made up of four members(type, layout, place, library), except for library, the other three members are related to Data. According to the principle of function: a single task, I suggest that `CastDataType` only handles the data type conversion on the same device.

File: paddle/framework/data_transform.cc [hunk 3]
```diff
@@ -23,5 +24,68 @@ DataTransformFnMap& DataTransformFnMap::Instance() {
...
+
+void TransDataLayout(const platform::DeviceContext* ctx,
+                     const KernelTypePair& kernel_pair, const Variable& in,
+                     Variable* out) {
+  PADDLE_ENFORCE(in.IsType<Tensor>(), "Only Support LoDTensor transform!.");
+  auto src = in.Get<Tensor>();
+  auto* dst = out->GetMutable<Tensor>();
+  dst->Resize(src.dims());
+  auto place = kernel_pair.second.place_;
+  CopyFrom(src, place, *ctx, dst);
```
Comment (chengduoZH) [inline paddle/framework/data_transform.cc]: Maybe `TransDataLayout` is wrong.
e.g.  `arr`'s layout is `NCHW` format, and data is:
```
    [[[[ 0,  1,  2],
         [ 3,  4,  5]],

        [[ 6,  7,  8],
         [ 9, 10, 11]]]]
```
Obviously, `arr`'s shape is (1,2,2,3).
If we change `NCHW` to `NHWC`, it's shape will be (1,3,2,2), and the data is:
```
    [[[[ 0,  6],
         [ 3,  9]],

        [[ 1,  7],
         [ 4, 10]],

        [[ 2,  8],
         [ 5, 11]]]]
```

File: paddle/framework/data_transform.cc [hunk 4]
```diff
@@ -23,5 +24,68 @@ DataTransformFnMap& DataTransformFnMap::Instance() {
...
+auto kernel_Layout1 = OpKernelType(proto::DataType::FP64, platform::CPUPlace(),
+                                   DataLayout::kNCHW, LibraryType::kPlain);
+
+void TransDataType(const platform::DeviceContext* ctx,
+                   const KernelTypePair& kernel_pair, const Variable& in,
+                   Variable* out) {
+  PADDLE_ENFORCE(in.IsType<Tensor>(), "Only Support Tensor transform!.");
+  auto src = in.Get<Tensor>();
+  auto* dst = out->GetMutable<Tensor>();
+  // TODO(dzhwinter): CPU <-> GPU need a copy here
```
[the remark this answers is not in the document]
Comment (dzhwinter) [reply paddle/framework/data_transform.cc]: Done.

File: paddle/framework/data_transform.cc [hunk 5]
```diff
@@ -23,5 +24,68 @@ DataTransformFnMap& DataTransformFnMap::Instance() {
...
+
+  auto dims = src.dims();
+  dst->Resize(dims);
+  auto dst_type = kernel_pair.second.data_type_;
+  auto src_type = kernel_pair.first.data_type_;
+
+  switch (src_type) {
```
[the remark this answers is not in the document]
Comment (dzhwinter) [reply paddle/framework/data_transform.cc]: No, this check should be done before the function.

File: paddle/framework/data_transform.cc (hunk 3 above)
Comment (dzhwinter) [reply paddle/framework/data_transform.cc]: Done.

File: paddle/framework/data_transform.cc [hunk 6]
```diff
@@ -23,5 +24,68 @@ DataTransformFnMap& DataTransformFnMap::Instance() {
...
+  CopyFrom(src, place, *ctx, dst);
+  dst->set_layout(kernel_pair.second.data_layout_);
+}
+
 }  // namespace framework
 }  // namespace paddle
+
+namespace f = paddle::framework;
+REGISTER_DATA_TRANSFORM_FN(f::kernel_FP32, f::kernel_FP64, f::TransDataType);
+REGISTER_DATA_TRANSFORM_FN(f::kernel_Layout0, f::kernel_Layout1,
```
[the remark this answers is not in the document]
Comment (dzhwinter) [reply paddle/framework/data_transform.cc]: Done.

File: paddle/framework/data_transform.h (hunk 1 above)
Comment (dzhwinter) [reply paddle/framework/data_transform.h]: Done.

File: paddle/framework/data_transform.h (hunk 2 above)
Comment (dzhwinter) [reply paddle/framework/data_transform.h]: Yes, we have re-designed the registrars.

Comment (dzhwinter) [reply paddle/framework/data_transform.h]: In the very beginning, we didn't figure out a method to do the multiple transforms.

For example, if we need to transform 0000 -> 1111, should we done in 4 step or 1 step?

>  *  For DataType, only test the FP32(float), FP64(double).
 *  e.g. 0000 -> FP32, CPUPlace, kNHWC, kPlain
 *       1111 -> FP64, GPUPlace, kNCHW, kMKLDNN
So this PR looks ugly when deal with CPU <-> GPU, and transform datatype the same time.

Now we have made the decision, we will do this in 4 step. ONE functor will only do its own duty, then we will only have four kinds of transform functor. We will achieve that in the final state.

File: paddle/framework/data_transform.cc [hunk 7]
```diff
@@ -23,5 +24,83 @@ DataTransformFnMap& DataTransformFnMap::Instance() {
...
+                     Variable* out) {
+  PADDLE_ENFORCE(in.IsType<Tensor>(), "Only Support Tensor transform!.");
+  PADDLE_ENFORCE(
+      platform::places_are_same_class(kernel_pair.first.place_,
+                                      kernel_pair.second.place_),
+      "TransDataType Only Support DataType transform on same place!");
+
+  auto src = in.Get<Tensor>();
+  auto* dst = out->GetMutable<Tensor>();
+  PADDLE_ENFORCE(arity(src.dims()) == 4, "Input Arity Only Suppport 4!");
```
Comment (chengduoZH) [inline paddle/framework/data_transform.cc]: `arity(src.dims())` can be replaced by `src.dims().size()`.  This is just a small problem.

File: paddle/framework/data_transform.h (hunk 2 above)
Comment (chengduoZH) [reply paddle/framework/data_transform.h]: For `LibraryType`, `Variable` does not need a transform functor, right?

File: paddle/framework/data_transform_test.cc [hunk 8]
```diff
@@ -83,17 +89,68 @@ TEST(DataTransform, Register) {
...
+
+  {
+    auto kernel1 = GenFromBit({1, 0, 0, 0});
+    auto kernel2 = GenFromBit({1, 0, 1, 0});
+    auto pair0 = std::make_pair(kernel1, kernel2);
+    instance.Get(pair0)(ctx, pair0, in, &out);
+  }
+
+  Tensor dst = out.Get<Tensor>();
+  EXPECT_TRUE(dst.layout() != src->layout());
```
Comment (chengduoZH) [inline paddle/framework/data_transform_test.cc]: `EXPECT_TRUE(dst.dims() == make_ddim({2, 2, 3, 1});`

File: paddle/framework/data_transform.h (hunk 2 above)
Comment (chengduoZH) [reply paddle/framework/data_transform.h]: For LibraryType, Variable does not need a transform functor, right?

Comment (dzhwinter) [reply paddle/framework/data_transform.h]: Yes, only 3 kinds of transform functor if you take it seriously.

File: paddle/framework/data_transform_test.cc (hunk 8 above)
Comment (dzhwinter) [reply paddle/framework/data_transform_test.cc]: good point, will fix the test in next PR.

**Pull request merged**